Nightly Trilinos failure: kokkos_swap update breakage

**Describe the bug**

Nightly Trilinos build (clang/15, serial backend so far) is failing to compile following merge of #6943 
The error pops up mixed in with kokkos-kernels/ode unit test output, but I did a grep search in kokkos-kernels and did not see usage of `kokkos_swap`...

```
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos-kernels/ode/unit_test/backends/Test_Serial_ODE.cpp:19:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos-kernels/test_common/Test_Serial.hpp:20:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Core.hpp:45:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Build/kokkos/KokkosCore_Config_DeclareBackend.hpp:22:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/decl/Kokkos_Declare_SERIAL.hpp:21:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Serial/Kokkos_Serial.hpp:37:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_TaskScheduler.hpp:34:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp:26:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Parallel.hpp:31:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_View.hpp:42:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_MinMax.hpp:21:
17:42:17 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Pair.hpp:31:
17:42:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Swap.hpp:43:19: error: no matching function for call to 'kokkos_swap'
17:42:17   static decltype(kokkos_swap(std::declval<T&>(), std::declval<T&>()))
17:42:17                   ^~~~~~~~~~~
17:42:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Array.hpp:138:13: note: in instantiation of template class 'Kokkos::Impl::is_swappable<const double>' requested here
17:42:17       Impl::is_swappable<T>::value>
17:42:17             ^
17:42:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos-kernels/ode/impl/KokkosODE_BDF_impl.hpp:313:40: note: in instantiation of template class 'Kokkos::Array<const double, 6>' requested here
17:42:17   const Kokkos::Array<const double, 6> alpha{
17:42:17                                        ^
17:42:17 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_LLVM1507_Serial_debug/Trilinos/kokkos/core/src/Kokkos_Swap.hpp:31:1: note: candidate template ignored: requirement 'std::is_move_assignable_v<const double>' was not satisfied [with T = const double]
17:42:17 kokkos_swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
```

**Please include the following for a minimal reproducer**

**Reproducer configuration** (blake):
```
git clone -b develop https://github.com/trilinos/Trilinos.git
TRILINOS_DIR=$PWD/Trilinos
git clone -b develop https://github.com/kokkos/kokkos.git
KOKKOS_DIR=$PWD/kokkos
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
KOKKOSKERNELS_DIR=$PWD/kokkos-kernels

cd $TRILINOS_DIR
ln -s ${KOKKOS_DIR}/kokkos kokkos
ln -s ${KOKKOSKERNELS_DIR}/kokkos-kernels kokkos-kernels
cd -

module load cmake llvm openblas/0.3.23

cmake \
  -D CMAKE_INSTALL_PREFIX="${TRILINOS_INSTALL_DIR}" \
  -D CMAKE_CXX_COMPILER="`which clang++`" \
  -D CMAKE_C_COMPILER="`which clang`" \
  -D CMAKE_Fortran_COMPILER="`which gfortran`" \
  -D CMAKE_CXX_FLAGS="-g -fPIC" \
  -D CMAKE_C_FLAGS="-g -fPIC" \
  -D CMAKE_Fortran_FLAGS="-g -fPIC" \
  -D CMAKE_BUILD_TYPE:STRING=DEBUG \
  -D BUILD_SHARED_LIBS:BOOL=OFF \
  -DTPL_ENABLE_MPI=OFF \
  -DTPL_ENABLE_BLAS:BOOL=ON \
   -DBLAS_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DBLAS_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_LAPACK:BOOL=ON \
   -DLAPACK_INCLUDE_DIRS:FILEPATH=${OPENBLAS_ROOT}/include \
   -DLAPACK_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DLAPACK_LIBRARY_NAMES:STRING="openblas" \
  -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
  -DTrilinos_ENABLE_TESTS=ON \
  -DTrilinos_MUST_FIND_ALL_TPL_LIBS=TRUE \
  -DTrilinos_ENABLE_COMPLEX=ON \
  -DTrilinos_ENABLE_OpenMP=OFF \
  -DTrilinos_ENABLE_Kokkos=ON \
  -D Kokkos_ENABLE_SERIAL=ON \
   -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
  -DTrilinos_ENABLE_KokkosKernels=ON \
   -D KokkosKernels_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Tpetra=ON \
   -D Tpetra_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Sacado=ON \
   -D Sacado_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Stokhos=ON \
   -D Stokhos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Intrepid2=ON \
   -D Intrepid2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Adelus=ON \
   -D Adelus_ENABLE_TESTS:BOOL=ON \
  -DTrilinos_ENABLE_Compadre=ON \
   -D Compadre_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Panzer=ON \
   -D Panzer_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Amesos2=ON \
   -D Amesos2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Zoltan2=ON \
   -D Zoltan2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Teuchos=ON \
   -D Teuchos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Ifpack2=ON \
   -D Ifpack2_ENABLE_TESTS=ON \
\
  -DTPL_ENABLE_Matio=OFF \
\
-DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
-DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR
```

--- a/core/src/Kokkos_Array.hpp
+++ b/core/src/Kokkos_Array.hpp
@@ -134,8 +134,9 @@ struct Array {
   }
 
  private:
+  template <class U = T>
   friend KOKKOS_INLINE_FUNCTION constexpr std::enable_if_t<
-      Impl::is_swappable<T>::value>
+      Impl::is_swappable<U>::value>
   kokkos_swap(Array<T, N>& a,
               Array<T, N>& b) noexcept(Impl::is_nothrow_swappable_v<T>) {
     for (std::size_t i = 0; i < N; ++i) {

--- a/core/src/Kokkos_Array.hpp
+++ b/core/src/Kokkos_Array.hpp
@@ -137,8 +137,13 @@ struct Array {
   template <class U = T>
   friend KOKKOS_INLINE_FUNCTION constexpr std::enable_if_t<
       Impl::is_swappable<U>::value>
-  kokkos_swap(Array<T, N>& a,
-              Array<T, N>& b) noexcept(Impl::is_nothrow_swappable_v<T>) {
+  kokkos_swap(
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+      Array<T, N, Proxy>& a, Array<T, N, Proxy>& b
+#else
+      Array<T, N>& a, Array<T, N>& b
+#endif
+      ) noexcept(Impl::is_nothrow_swappable_v<T>) {
     for (std::size_t i = 0; i < N; ++i) {
       kokkos_swap(a[i], b[i]);
     }

--- a/core/src/Kokkos_Array.hpp
+++ b/core/src/Kokkos_Array.hpp
@@ -137,13 +137,8 @@ struct Array {
   template <class U = T>
   friend KOKKOS_INLINE_FUNCTION constexpr std::enable_if_t<
       Impl::is_swappable<U>::value>
-  kokkos_swap(
-#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
-      Array<T, N, Proxy>& a, Array<T, N, Proxy>& b
-#else
-      Array<T, N>& a, Array<T, N>& b
-#endif
-      ) noexcept(Impl::is_nothrow_swappable_v<T>) {
+  kokkos_swap(Array<T, N>& a,
+              Array<T, N>& b) noexcept(Impl::is_nothrow_swappable_v<T>) {
     for (std::size_t i = 0; i < N; ++i) {
       kokkos_swap(a[i], b[i]);
     }

--- a/core/src/Kokkos_Swap.hpp
+++ b/core/src/Kokkos_Swap.hpp
@@ -37,6 +37,26 @@ kokkos_swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
 
 namespace Impl {
 
+// Workaround for the definition of is_nothrow_swappable_v
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1140)
+template <class T>
+struct is_swappable {
+  template <class U>
+  static decltype(kokkos_swap(std::declval<T&>(), std::declval<T&>()))
+  test_swap(int) noexcept(noexcept(kokkos_swap(std::declval<T&>(),
+                                               std::declval<T&>())));
+  struct Nope {};  // test_swap must return a complete type for the definition
+                   // of nothrow below
+  template <class U>
+  static Nope test_swap(long);
+  static constexpr bool value =
+      !std::is_same_v<decltype(test_swap<T>(0)), Nope>;
+  static constexpr bool nothrow = noexcept(test_swap<T>(0));
+};
+
+template <class T>
+inline constexpr bool is_nothrow_swappable_v = is_swappable<T>::nothrow;
+#else
 template <class T>
 struct is_swappable {
   template <class U>
@@ -52,6 +72,7 @@ struct is_swappable {
 template <class T>
 inline constexpr bool is_nothrow_swappable_v =
     noexcept(kokkos_swap(std::declval<T&>(), std::declval<T&>()));
+#endif
 
 }  // namespace Impl
 

--- a/core/src/Kokkos_Swap.hpp
+++ b/core/src/Kokkos_Swap.hpp
@@ -37,26 +37,6 @@ kokkos_swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
 
 namespace Impl {
 
-// Workaround for the definition of is_nothrow_swappable_v
-#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1140)
-template <class T>
-struct is_swappable {
-  template <class U>
-  static decltype(kokkos_swap(std::declval<T&>(), std::declval<T&>()))
-  test_swap(int) noexcept(noexcept(kokkos_swap(std::declval<T&>(),
-                                               std::declval<T&>())));
-  struct Nope {};  // test_swap must return a complete type for the definition
-                   // of nothrow below
-  template <class U>
-  static Nope test_swap(long);
-  static constexpr bool value =
-      !std::is_same_v<decltype(test_swap<T>(0)), Nope>;
-  static constexpr bool nothrow = noexcept(test_swap<T>(0));
-};
-
-template <class T>
-inline constexpr bool is_nothrow_swappable_v = is_swappable<T>::nothrow;
-#else
 template <class T>
 struct is_swappable {
   template <class U>
@@ -69,6 +49,13 @@ struct is_swappable {
       !std::is_same_v<decltype(test_swap<T>(0)), Nope>;
 };
 
+#if defined(KOKKOS_COMPILER_NVCC) && (KOKKOS_COMPILER_NVCC < 1140)
+template <class T>
+inline constexpr bool is_nothrow_swappable_v =
+    is_swappable<T>::value&& noexcept(
+        kokkos_swap(std::declval<std::remove_const_t<T>&>(),
+                    std::declval<std::remove_const_t<T>&>()));
+#else
 template <class T>
 inline constexpr bool is_nothrow_swappable_v =
     noexcept(kokkos_swap(std::declval<T&>(), std::declval<T&>()));

--- a/core/unit_test/TestArray.cpp
+++ b/core/unit_test/TestArray.cpp
@@ -120,6 +120,13 @@ static_assert(test_array_aggregate_initialization());
   }
 }
 
+constexpr bool test_array_const_qualified_element_type() {
+  Kokkos::Array<int const, 1> a{255};
+  return a[0] == 255;
+}
+
+static_assert(test_array_const_qualified_element_type());
+
 // User-defined type providing a sepcialization of kokkos_swap
 struct MyInt {
   int i;

